crates/profiling/src/vtune_linux.rs

```diff
@@ -0,0 +1,138 @@
+    method_id: HashMap<(usize, DefinedFuncIndex), u32>,
+}
+
+impl VTuneAgent {
+    /// Intialize a VTuneAgent and write out the header
+    pub fn new() -> Result<Self> {
+        let state = State {
+            method_id: HashMap::new(),
+        };
+        Ok(VTuneAgent {
+            state: Mutex::new(state),
+        })
+    }
+}
+
+impl State {
+    /// Return the unique method ID for use with the ittapi
+    pub fn get_method_id(&mut self, module_id: usize, func_idx: DefinedFuncIndex) -> u32 {
+        let method_id;
+        unsafe {
+            method_id = self
+                .method_id
+                .entry((module_id, func_idx))
+                .or_insert_with(|| iJIT_GetNewMethodID());
+        }
+        *method_id
+    }
+
+    /// Load module
+    pub fn event_load(
+        &mut self,
+        method_id: u32,
+        module_name: &str,
+        method_name: &str,
+        addr: *const u8,
+        len: usize,
+    ) -> () {
+        let filename = "unknown_filename";
+        let jmethod = _iJIT_Method_Load {
+            method_id: method_id,
+            method_name: CString::new(method_name)
+                .expect("CString::new failed")
+                .into_raw(),
+            method_load_address: addr as *mut ::std::os::raw::c_void,
+            method_size: len as u32,
+            line_number_size: 0,
+            line_number_table: ptr::null_mut(),
+            class_id: 0,
+            class_file_name: CString::new(module_name)
+                .expect("CString::new failed")
+                .into_raw(),
+            source_file_name: CString::new(filename)
+                .expect("CString::new failed")
+                .into_raw(),
+        };
+        let jmethod_ptr = Box::into_raw(Box::new(jmethod));
+        unsafe {
+            let _ret = iJIT_NotifyEvent(
+                iJIT_jvm_event_iJVM_EVENT_TYPE_METHOD_LOAD_FINISHED,
+                jmethod_ptr as *mut ::std::os::raw::c_void,
+            );
+        }
+    }
+
+    /// Shutdown module
+    pub fn _event_shutdown(&mut self) -> () {
+        unsafe {
+            let _ret = iJIT_NotifyEvent(iJIT_jvm_event_iJVM_EVENT_TYPE_SHUTDOWN, ptr::null_mut());
+        }
+    }
+}
+
+use crate::ProfilingAgent;
+impl ProfilingAgent for VTuneAgent {
+    fn module_load(
+        &self,
+        module: &Module,
+        functions: &PrimaryMap<DefinedFuncIndex, *mut [VMFunctionBody]>,
+        dbg_image: Option<&[u8]>,
+    ) {
+        self.state
+            .lock()
+            .unwrap()
+            .module_load(module, functions, dbg_image);
+    }
+}
+
+impl State {
+    fn module_load(
+        &mut self,
+        module: &Module,
+        functions: &PrimaryMap<DefinedFuncIndex, *mut [VMFunctionBody]>,
+        _dbg_image: Option<&[u8]>,
+    ) -> () {
+        for (idx, func) in functions.iter() {
+            let (addr, len) = unsafe { ((**func).as_ptr() as *const u8, (**func).len()) };
+            let module_name = module.name.as_ref().unwrap();
```

Comment:
This `.unwrap()` is likely to panic since I don't think that all wasm modules have names associated with them